internal/core/src/segcore/SegmentSealed.h

```diff
@@ -73,63 +83,69 @@ class SegmentSealed : public SegmentInternalInterface {
         return SegmentType::Sealed;
     }
 
-    PinWrapper<const index::IndexBase*>
-    chunk_index_impl(FieldId field_id,
-                     std::string path,
-                     int64_t chunk_id) const override {
-        JSONIndexKey key;
-        key.field_id = field_id;
-        key.nested_path = path;
-        AssertInfo(json_indexings_.find(key) != json_indexings_.end(),
-                   "Cannot find json index with path: " + path);
-        return PinWrapper<const index::IndexBase*>(
-            json_indexings_.at(key).get());
-    }
-
     virtual bool
     HasIndex(FieldId field_id) const override = 0;
     bool
     HasIndex(FieldId field_id,
              const std::string& path,
              DataType data_type,
              bool any_type = false) const override {
-        JSONIndexKey key;
-        key.field_id = field_id;
-        key.nested_path = path;
-        auto index = json_indexings_.find(key);
-        if (index == json_indexings_.end()) {
-            return false;
-        }
-        if (any_type) {
-            return true;
-        }
-        return index->second->IsDataTypeSupported(data_type);
+        auto it = std::find_if(
+            json_indices.begin(),
+            json_indices.end(),
+            [field_id, path, data_type, any_type](const JsonIndex& index) {
+                if (index.field_id != field_id) {
+                    return false;
+                }
+                if (index.cast_type.data_type() ==
+                    JsonCastType::DataType::JSON) {
+                    // for json flat index, path should be a subpath of nested_path
+                    return path.substr(0, index.nested_path.length()) ==
+                           index.nested_path;
+                }
+                if (any_type) {
+                    return true;
+                }
+                return index.nested_path == path &&
+                       index.index->IsDataTypeSupported(data_type);
+            });
+        return it != json_indices.end();
     }
 
  protected:
-    struct JSONIndexKey {
+    virtual PinWrapper<const index::IndexBase*>
+    chunk_index_impl(FieldId field_id, int64_t chunk_id) const override = 0;
+
+    PinWrapper<const index::IndexBase*>
+    chunk_index_impl(FieldId field_id,
```

Comment:
how about renaming this method to chunk_json_index_impl? would make it easier to read code

Reply:
question: through code searching, chunk_scalar_index is also calling this method, what is the difference between chunk_scalar_index and chunk_json_index?

Reply:
The return type of chunk_scalar_index must be specified with the index type, and the reason why we need chunk_json_index is already explained in the comments.